keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota. This could happen when the job corresponding to these workloads don't
+specify priorities or specify the same priority.
+
+My use case be supported by adding `kueue.x-k8s.io/startup-requirements: QuotaAndPodsReady`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+To mitigate this issue we introduce introduce a timeout on reaching the `PodsReady`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsReady condition](#timeout-on-reaching-the-podsready-condition)).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsReady workload condition
+
+First, we introduce a new workload condition, called `PodsReady`, to indicate
+that the first batch of a workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+
+The `PodsReady` condition is addded to the workload by the Kueue's Job
+reconciler in reaction to a status update of the corresponding Job. Note that,
+verifying if the condition should be added does not require an extra API call as
+the Kueue's Job reconciller already fetches the latest Job object at the
+beginning of the `Reconcile` function. Also note, there is no new condition
+introduced to the Job API.
+
+### Workload API
+
+We extend the Workload API with the new `startupRequirements` field to
+enable the new behavior.
+
+```golang
+// StartupRequirements describes workload startup requirements
+type StartupRequirements string
+
+const (
+  // QuotaAndPodsReady means that workload requires all pods to be running at
+  // the same time
+  StartupRequirementsQuotaAndPodsReady StartupRequirements = "QuotaAndPodsReady"
```

Comment:
StartRequirements? (shorter, simpler)

Reply:
Out of StartupRequirements vs StartRequirements I prefer StartupRequirements which reflects more on time and effort to start "If something takes some time to get fully started, start up can have that implication too." https://ell.stackexchange.com/questions/104319/start-vs-start-up. 

Also, technically the workload gets started as soon as the Job is unsuspended, so maybe Start/ Startup is not best after all. How about: `RunRequirements` or `ProgressRequirements`?

Reply:
`Start` is good. It matches the `StartTime` of the Job.

Reply:
Wouldn't actually be misleading, because `StartTime` in job points to the moment of unsuspending, whereas the `StartupRequirements` , but QuotaAndPodsReady value point to the moment when all pods are ready.

Reply:
So `StartupRequirements` would mark the distinction actually or we can go further and name it `RunningRequirements`,

Reply:
It's about the requirements we have before starting the job. But sure, in the first implementation we will also block after we already started the Job.

I don't have a strong opinion. Both `Start` and `Startup` are ok with me.

Reply:
I'm also ok with `Running`